### ISSUE TITLE
class imbalance in graph classification

### ISSUE THREAD

**Issue by jihwanp** [pyg-team/pytorch_geometric#2089]
## ❓ Questions & Help

Hello. I'd like to use custom data set in graph classification. Is there any tools for solvinig the class imbalance? like oversampling or focal loss.

**Comment by wsad1** [comment]
1.`torch.nn.NLLLoss` and `torch.nn.CrossEntropyLoss` have a `weight` argument. So you could give a higher weight to the class with fewer samples. This might work similar to oversampling.
2. I couldn't find any inbuilt implementation of focal loss in torch. But you could implement a custom loss function like so.
```
def focal_loss(pred_proba, target, weight=None, gamma=2):
    
    #reduction='none'. No aggregation performed.
    nll = NLLLoss(weight=weight,reduction='none')
    loss = nll(pred_proba,target)#shape=N
    class_preds = pred_proba.gather(dim=1,index=target.unsqueeze(1)).squeeze(1)
    
    return torch.mean(torch.pow(1-class_preds, gamma) * loss)
```
Note: the custom loss function should only use pytorch operations (https://discuss.pytorch.org/t/custom-loss-functions/29387)

**Comment by jihwanp** [comment]
> 1.`torch.nn.NLLLoss` and `torch.nn.CrossEntropyLoss` have a `weight` argument. So you could give a higher weight to the class with fewer samples. This should work similar to oversampling.
> 2. I couldn't find any inbuilt implementation of focal loss in torch. But you could implement a custom loss function like so.
> 
> ```
> def focal_loss(pred_proba, target, weight=None, gamma=2):
>     
>     #reduction='none'. No aggregation performed.
>     nll = NLLLoss(weight=weight,reduction='none')
>     loss = nll(pred_proba,target)#shape=N
>     class_preds = pred_proba.gather(dim=1,index=target.unsqueeze(1)).squeeze(1)
>     
>     return torch.mean(torch.pow(1-class_preds, gamma) * loss)
> ```
> 
> Note: custom loss functions should only use pytorch operations (https://discuss.pytorch.org/t/custom-loss-functions/29387)

Oh I forgot nll_loss offers weight argument. Thanks for the information!

**Comment by rusty1s** [comment]
Yes, utilizing the `weight` in `nll_loss` is a good idea. In case you want to ensure balanced data sampling, you can provide your own `sampler` for data loading, see [here](https://discuss.pytorch.org/t/balanced-sampling-between-classes-with-torchvision-dataloader/2703/44).
